ui/src/app/pages/data/concept/concept-homepage.tsx

```diff
@@ -310,14 +318,69 @@ export const ConceptHomepage = withCurrentWorkspace()(
       }
     }
 
+    validateInputForMySQL() {
+      const {currentInputString} = this.state;
+      const inputErrors = new Set(); // use Set to prevent duplicate messages
+      let openParensCount = 0;
+      let unclosedQuotes = false;
+      for (let i = 0; i < currentInputString.length; i++) {
+        const character = currentInputString[i];
+        if (character === '"') {
+          unclosedQuotes = !unclosedQuotes;
+          continue;
+        }
+        if (unclosedQuotes) {
+          // inside a quote, no need to validate further
+          continue;
+        }
+        // Check for characters that break search
+        if ('~@[]|<>'.indexOf(character) > -1) {
+          inputErrors.add('The following characters are not allowed in the search string: ~ @ [ ] | < >');
+          continue;
+        }
+        // Check for trailing + or -
+        if ('+-'.indexOf(character) > -1 && (currentInputString[i + 1] === ' ' || currentInputString[i + 1] === undefined)) {
+          inputErrors.add(`Trailing ${character} characters are not allowed in the search string`);
+          continue;
+        }
+        const parenPosition = '()'.indexOf(character);
+        if (parenPosition === -1) {
+          // Parens are the last character check, so we can continue if it's something else
+          continue;
+        }
+        if (parenPosition === 0) {
+          openParensCount++; // increment the number of unclosed parens
+        } else {
+          if (openParensCount === 0) {
+            // too many closing parens
+            inputErrors.add('There are too many ) characters in the search string');
+            continue;
+          }
+          openParensCount--; // decrement the number of unclosed parens
+        }
+      }
+      if (openParensCount > 0) {
+        // unclosed paren
+        inputErrors.add('There is an unclosed ( in the search string');
+      }
+      if (unclosedQuotes) {
+        // unclosed quote
+        inputErrors.add('There is an unclosed " in the search string');
+      }
+      this.setState({inputErrors: Array.from(inputErrors)});
+      return inputErrors.size === 0; // return true if no errors
+    }
+
```

Comment:
I guess since you have the ability to embed `"` and `( )` you have to parse each character instead of using regex. I know this is a product decision on unifying this search with CB and DSB, but we should make this reusable by other components.

Reply:
AFAIK there's no way to detect an imbalance of quotes or parens using regex. The [SO solution](https://stackoverflow.com/questions/26507087/escape-string-for-use-in-mysql-fulltext-search) we discussed seems to just strip out anything except letters, numbers and underscores. 

Yeah, whenever we unify the tools, will definitely make this (or whatever validation we end up using) reusable.

Reply:
I think we just need to write up what all apps currently do and then present it to the PMs and see if we can get consensus